client/shipping-label/state/actions.js

```diff
@@ -70,18 +99,29 @@ export const purchaseLabel = () => ( dispatch, getState, { callbackURL, nonce, s
 		}
 	};
 
-	const form = getState().shippingLabel.form;
-	const formData = {
-		origin: form.origin.values,
-		destination: form.destination.values,
-		payment_method_id: '123456789',
-		carrier: 'usps',
-		packages: form.packages.values.map( ( pckg, index ) => ( {
-			...omit( pckg, [ 'items', 'id' ] ),
-			service_id: form.rates.values[ index ],
-			products: flatten( pckg.items.map( ( item ) => fill( new Array( item.quantity ), item.product_id ) ) ),
-		} ) ),
-	};
+	let form = getState().shippingLabel.form;
+	const addressesValidation = [];
+	if ( ! form.origin.isValidated ) {
+		addressesValidation.push( validateAddress( dispatch, form.origin.values, 'origin', addressValidationURL, nonce ) );
+	}
+	if ( ! form.destination.isValidated ) {
+		addressesValidation.push( validateAddress( dispatch, form.destination.values, 'destination', addressValidationURL, nonce ) );
+	}
+
+	Promise.all( addressesValidation ).then( () => {
+		form = getState().shippingLabel.form;
+		const formData = {
+			origin: form.origin.pickNormalized ? form.origin.normalized : form.origin.values,
+			destination: form.destination.pickNormalized ? form.destination.normalized : form.destination.values,
+			payment_method_id: '123456789',
+			carrier: 'usps',
+			packages: form.packages.values.map( ( pckg, index ) => ( {
+				...omit( pckg, [ 'items', 'id' ] ),
+				service_id: form.rates.values[ index ],
+				products: flatten( pckg.items.map( ( item ) => fill( new Array( item.quantity ), item.product_id ) ) ),
+			} ) ),
+		};
 
-	saveForm( setIsSaving, setSuccess, noop, setError, callbackURL, nonce, submitMethod, formData );
+		saveForm( setIsSaving, setSuccess, noop, setError, purchaseURL, nonce, 'POST', formData );
+	} ).catch( noop );
```

Comment:
In Chrome, an unhandled rejected Promise spits an error in the console, so if we want to use Promises for this I have to stick this ugly workaround. `node.js` follows the Promise spec by the book, and silently ignores any unhandled rejection, so this particular behaviour surprised me (and annoyed me).

I considered **not** using Promises for this, but converting this `Promise.all(validations)` to something with callbacks would be an uglier mess.